### ISSUE TITLE
UNION ALL

### ISSUE THREAD

**Issue by murphydan** [coldbox-modules/qb#3]
Need to be able to use UNION ALL to combine two or more datasets.

**Comment by dswitzer** [comment]
I'm hoping to post an official pull request for this soon, but I'm using a fork of the code that runs under CF10 at the moment and need to get an environment set up where I can run the original source before I can do that.

Listed below are the changes required to add support for union() and unionAll() methods to QB.

The `union()` statements do need to be at the _end_ of QB statements, because any bindings are merged into the main query. I thought about doing the merging in the Grammar layer so the union methods could be in any order, but that means each time the toSQL() method is called, the bindings would be rebound. We could just track if the bindings have already been added and skip adding them again, but it seemed like it didn't add much value.  If the order needs to be agnostic, then we'd probably need to change the getBindings() statement to extract the bindings from the stored union QB instances.


### QueryBuilder.cfc

Add the following new property between "havings" and "orders":

```
    /** 
    * An array of UNION statements.
    * @default []
    */
    property name="unions" type="array";
```
Add the following to the `setDefaultValues()` method:

```
        variables.unions = [];
```

Add the following method after the `addBindings()` function:

```
    /** 
    * Merges bindings from a derived/sub-query
    *
    * @query The query to merge the bindings.
    *
    * @return qb.models.Query.QueryBuilder
    */
    public QueryBuilder function mergeBindings(
        required QueryBuilder input
    ) {

    		var bindings = input.getRawBindings();

    		for( var type in variables.bindings ){
    			arrayAppend(variables.bindings[ type ], bindings[ type ], true);
    		}

    		return this;
    }
```

Add the following method after the `orderBy()` function:

```
    /*******************************************************************************\
    |         UNION/CTE functions                                                   |
    \*******************************************************************************/

    /**
    * Add a UNION statement to the SQL.
    *
    * @input    Either a QueryBuilder instance or a closure to define the derived query.
    * @all Determines if UNION statement should be a "UNION ALL".  Passing this as an argument is discouraged.  Use the dedicated `unionAll` where possible.
    * @return qb.models.Query.QueryBuilder
    */
    public QueryBuilder function union(required any input, boolean all=false) {
        var qb = arguments.input;

        // since we have a callback, we generate a new query object and pass it into the callback
        if( isClosure(qb) ){
            var subquery = newQuery();
            qb(subquery);
            // replace the original query builder with the results of the sub-query
            qb = subquery;
        }

        // track the union statement
        arrayAppend(variables.unions, {query=qb, all=arguments.all});

       	// we need to merge the bindings of the query
       	mergeBindings(qb);

        return this;
    }

    /**
    * Add a UNION ALL statement to the SQL.
    *
    * @input Either a QueryBuilder instance or a closure to define the derived query.
    * @return qb.models.Query.QueryBuilder
    */
    public QueryBuilder function unionAll(required any input) {
        return union(arguments.input, true);
    }
```


### BaseGrammar.cfc

Change the `selectComponents` to:

```
    /**
    * The different components of a select statement in the order of compilation.
    */
    variables.selectComponents = [
        "aggregate", "columns", "from", "joins", "wheres",
        "groups", "havings", "unions", "orders", "limitValue", "offsetValue"
    ];
```

Add the following method after the `compileHaving()` function:

```
    /** MOD: New method
    * Compiles the UNION portion of a sql statement.
    *
    * @query The Builder instance.
    * @orders The union clauses.
    *
    * @return string
    */
    private string function compileUnions( required QueryBuilder query, required array unions ) {
        if ( arrayIsEmpty(arguments.unions) ) {
            return "";
        }
        var sql = utils.arrayMap(arguments.unions, function (union){
           var sql = arguments.union.query.toSQL();

            if( arrayLen(arguments.union.query.getOrders()) ){
                throw(
                    type = "OrderByNotAllowed",
                    message = "The ORDER BY clause is not allowed in a UNION statement.",
                    detail = "A QueryBuilder instance used in a UNION statement is cannot have any ORDER BY clause, as this is not allowed by SQL. Only the outer most query is allowed to specify an ORDER BY clause which will be used on the unioned queries."
                );
            }

            return "UNION " & (arguments.union.all ? "ALL " : "") & sql;
        });

        return trim( arrayToList(sql, ' ') );
    }
```

### MSSQLGrammar.cfc

Change the `selectComponents` to:

```
    /**
    * The different components of a select statement in the order of compilation.
    */
    variables.selectComponents = [
        "aggregate", "columns", "from", "joins", "wheres",
        "groups", "havings", "unions", "orders", "offsetValue", "limitValue"
    ];
```

**Comment by elpete** [comment]
This is looking great!  I can't wait to see your PR.  Please include tests, as well. 👍

**Comment by dswitzer** [comment]
I'm working on support for Common Table Expressions as well.

**Comment by dswitzer** [comment]
I actually tweaked the code to track bindings for UNION clauses separately. 

All I needed to do was:

* Add a `"union" = []` to the `variables.bindings`
* Instead of the `mergeBindings()` (the function is no longer needed for this feature), use: `addBindings( qb.getBindings(), "union" );`
* In the `getBindings()` function, add `"union"` at the end of the `bindingOrder` array.
* In the `clearBindings()` function, add the `"union"` after the `"where"` item in the `arguments.only` declaration.

This should make the code so that the order of the `union` statements no longer matter.
